imap/pushsub_db.c

```diff
@@ -0,0 +1,449 @@
+        { NULL,             SQLITE_NULL,    { .s = NULL                 } } };
+    const char *cmd;
+    int r;
+
+    if (psdata->rowid) {
+        cmd = CMD_UPDATE;
+
+        bval[5].name = ":rowid";
+        bval[5].type = SQLITE_INTEGER;
+        bval[5].val.i = psdata->rowid;
+    }
+    else {
+        cmd = CMD_INSERT;
+
+        bval[5].name = ":mailbox";
+        bval[5].type = SQLITE_TEXT;
+        bval[5].val.s = psdata->mailbox;
+        bval[6].name = ":id";
+        bval[6].type = SQLITE_TEXT;
+        bval[6].val.s = psdata->id;
+    }
+
+    r = sqldb_exec(pushsubdb->db, cmd, bval, NULL, NULL);
+
+    return r;
+}
+
+
+#define CMD_DELETE "DELETE FROM push_subscriptions WHERE rowid = :rowid;"
+
+EXPORTED int pushsubdb_delete(struct pushsub_db *pushsubdb, unsigned rowid)
+{
+    struct sqldb_bindval bval[] = {
+        { ":rowid", SQLITE_INTEGER, { .i = rowid } },
+        { NULL,     SQLITE_NULL,    { .s = NULL  } } };
+
+    return sqldb_exec(pushsubdb->db, CMD_DELETE, bval, NULL, NULL);
+}
+
+EXPORTED int pushsub_ensure_folder(const char *userid, struct mailbox **mailboxptr)
+{
+    struct mboxlock *namespacelock = NULL;
+    char *mboxname = pushsub_mboxname(userid);
+    int r = mboxlist_lookup(mboxname, NULL, NULL);
+
+    if (r == IMAP_MAILBOX_NONEXISTENT) {
+        namespacelock = user_namespacelock(userid);
+
+        if (!namespacelock) {
+            r = IMAP_MAILBOX_LOCKED;
+            goto done;
+        }
+
+        /* maybe we lost the race on this one */
+        r = mboxlist_lookup(mboxname, NULL, NULL);
+    }
+
+    if (!r && mailboxptr) {
+        r = mailbox_open_iwl(mboxname, mailboxptr);
+        if (r) {
+            syslog(LOG_ERR, "IOERROR: failed to open %s (%s)",
+                   mboxname, error_message(r));
+            goto done;
+        }
+    }
+
+    else if (r == IMAP_MAILBOX_NONEXISTENT) {
+        /* Create locally */
+        struct mailbox *mailbox = NULL;
+        mbentry_t mbentry = MBENTRY_INITIALIZER;
+        mbentry.name = (char *) mboxname;
+        mbentry.mbtype = MBTYPE_JMAPPUSHSUB;
+
+        r = mboxlist_createmailbox(&mbentry, 0/*options*/, 0/*highestmodseq*/,
+                                   1/*isadmin*/, userid, NULL/*auth_state*/,
+                                   0/*flags*/, &mailbox);
+        if (r) {
+            syslog(LOG_ERR, "IOERROR: failed to create %s (%s)",
+                   mboxname, error_message(r));
+            goto done;
+        }
+
+        if (mailboxptr) *mailboxptr = mailbox;
+        else mailbox_close(&mailbox);
+    }
+
+  done:
+    mboxname_release(&namespacelock);
+    free(mboxname);
+    return r;
+}
+
+EXPORTED char *pushsub_mboxname(const char *userid)
+{
+    struct buf boxbuf = BUF_INITIALIZER;
+    char *res = NULL;
+
+    init_internal();
+
+    buf_setcstr(&boxbuf, config_getstring(IMAPOPT_JMAPPUSHSUBSCRIPTIONFOLDER));
+
+    res = mboxname_user_mbox(userid, buf_cstring(&boxbuf));
```

Comment:
Could this be something like `res = mboxname_user_mbox(userid, config_getstring(...));`?  I don't see what the struct buf gets us here.  Maybe it's leftover from an earlier implementation?